contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,712 @@
+        error |= pthread_mutex_destroy(&ctx->jobWrite_mutex);
+        error |= pthread_cond_destroy(&ctx->jobWrite_cond);
+        error |=  (ctx->dstFile != NULL && ctx->dstFile != stdout) ? fclose(ctx->dstFile) : 0;
+        error |= ZSTD_isError(ZSTD_freeCCtx(ctx->cctx));
+        free(ctx->input.buffer.start);
+        if (ctx->jobs){
+            freeCompressionJobs(ctx);
+            free(ctx->jobs);
+        }
+        free(ctx);
+        return error;
+    }
+}
+
+static adaptCCtx* createCCtx(unsigned numJobs, const char* const outFilename)
+{
+
+    adaptCCtx* ctx = calloc(1, sizeof(adaptCCtx));
+    if (ctx == NULL) {
+        DISPLAY("Error: could not allocate space for context\n");
+        return NULL;
+    }
+    ctx->compressionLevel = g_compressionLevel;
+    pthread_mutex_init(&ctx->jobCompressed_mutex, NULL);
+    pthread_cond_init(&ctx->jobCompressed_cond, NULL);
+    pthread_mutex_init(&ctx->jobReady_mutex, NULL);
+    pthread_cond_init(&ctx->jobReady_cond, NULL);
+    pthread_mutex_init(&ctx->allJobsCompleted_mutex, NULL);
+    pthread_cond_init(&ctx->allJobsCompleted_cond, NULL);
+    pthread_mutex_init(&ctx->jobWrite_mutex, NULL);
+    pthread_cond_init(&ctx->jobWrite_cond, NULL);
+    ctx->numJobs = numJobs;
+    ctx->jobReadyID = 0;
+    ctx->jobCompressedID = 0;
+    ctx->jobWriteID = 0;
+    ctx->lastDictSize = 0;
+    ctx->jobs = calloc(1, numJobs*sizeof(jobDescription));
+    /* initializing jobs */
+    {
+        unsigned jobNum;
+        for (jobNum=0; jobNum<numJobs; jobNum++) {
+            jobDescription* job = &ctx->jobs[jobNum];
+            job->src.start = malloc(2 * FILE_CHUNK_SIZE);
+            job->dst.start = malloc(ZSTD_compressBound(FILE_CHUNK_SIZE));
+            job->lastJob = 0;
+            if (!job->src.start || !job->dst.start) {
+                DISPLAY("Could not allocate buffers for jobs\n");
+                freeCCtx(ctx);
+                return NULL;
+            }
+            job->src.capacity = FILE_CHUNK_SIZE;
+            job->dst.capacity = ZSTD_compressBound(FILE_CHUNK_SIZE);
+        }
+    }
+    ctx->nextJobID = 0;
+    ctx->threadError = 0;
+    ctx->allJobsCompleted = 0;
+    ctx->adaptParam = DEFAULT_ADAPT_PARAM;
+    ctx->cctx = ZSTD_createCCtx();
+    ctx->input.filled = 0;
+    ctx->input.buffer.capacity = 2 * FILE_CHUNK_SIZE;
+    ctx->input.buffer.start = malloc(ctx->input.buffer.capacity);
+    if (!ctx->input.buffer.start) {
+        DISPLAY("Error: could not allocate input buffer\n");
+        freeCCtx(ctx);
+        return NULL;
+    }
+    if (!ctx->cctx) {
+        DISPLAY("Error: could not allocate ZSTD_CCtx\n");
+        freeCCtx(ctx);
+        return NULL;
+    }
+    if (!ctx->jobs) {
+        DISPLAY("Error: could not allocate space for jobs during context creation\n");
+        freeCCtx(ctx);
+        return NULL;
+    }
+    {
+        unsigned const stdoutUsed = !strcmp(outFilename, stdoutmark);
```

Comment:
General note : it feels a bit strange to have a `FILE*` object opened inside `adaptCCtx`.
It seems to imply both lifetimes are tied.
I suspect that's because you have a thread dedicated to write into `dstFile`.

I believe `adaptCCtx` can be more general, as the speed adaptation method doesn't depend on any specific signal from `FILE* dstFile`.

That being said, for a demo, it's not yet a big deal.
Just food for thought, to keep in mind for an improved version with the aim to become part of `zstd` cli or library.